[rust] Test Selenium Manager on Linux arm64

dennisameling opened the pull request:
### **User description**
<!-- Thanks for Contributing to Selenium! -->
<!-- Please read our contribution guidelines: https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md -->

### 🔗 Related Issues

https://github.com/SeleniumHQ/selenium/issues/15801

<!-- Example: Fixes #1234 or Closes #5678 -->
<!-- If the reason for this PR is not obvious, consider creating an issue for it first -->

### 💥 What does this PR do?
<!-- Describe what this change includes and how it works -->

This PR ensures that the Selenium Manager test suite passes on Linux arm64, and enables CI tests for this platform through GitHub's [recently released Linux arm64 runners](https://github.blog/changelog/2025-01-16-linux-arm64-hosted-runners-now-available-for-free-in-public-repositories-public-preview/).

### 🔧 Implementation Notes

I'm new to Rust, but wanted Selenium Manager to explicitly fail if users try to run it for Chrome or Edge on Linux arm64, since it's not supported. Previously, the code would silently download `linux64` binaries which are for Linux x64, and cause segfaults on Linux arm64.

The setup I went for at least ensures that Firefox/Geckodriver on Linux arm64 will work and are tested properly as well.

<!--- Why did you implement it this way? -->
<!--- What alternatives to this approach did you consider? -->

### 💡 Additional Considerations
<!--- Are there any decisions that need to be made before accepting this PR? -->
<!--- Is there any follow-on work that needs to be done? (e.g., docs, tests, etc.) -->

- After merging this PR, additional work has to be done to actually release the Linux arm64 binary for Selenium Manager, and include it in the main Selenium packages somehow.
- The docs at https://www.selenium.dev/documentation/selenium_manager/#alternative-architectures will need to be updated if the team is open to merging this PR. I'm happy to open a separate PR for that if desired.

### 🔄 Types of changes
<!-- ✂️ Please delete anything that doesn't apply -->
- New feature (non-breaking change which adds functionality *and tests!*)


___

### **PR Type**
Enhancement


___

### **Description**
- Add Linux ARM64 support for Selenium Manager

- Enable CI testing on GitHub's ARM64 runners

- Restrict Chrome/Edge to Firefox-only on ARM64

- Update test suite for ARM64 compatibility


___

### **Changes diagram**

```mermaid
flowchart LR
  A["Linux ARM64 Detection"] --> B["Browser Support Check"]
  B --> C["Firefox: Supported"]
  B --> D["Chrome/Edge: Error"]
  E["Test Suite Updates"] --> F["ARM64 Conditional Logic"]
  G["CI Configuration"] --> H["GitHub ARM64 Runners"]
```


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Error handling</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>chrome.rs</strong><dd><code>Add ARM64 unsupported error for Chrome</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16045/files#diff-5ac161093a229c7c74c75c64f3be31a337c109d1ff11fcb0efe7bcc94d6e77be">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>edge.rs</strong><dd><code>Add ARM64 unsupported error for Edge</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16045/files#diff-dcd62c3bd133281f5bd97d8a876bcf753d7a4f3edd6ccbacf44eff3e9fcc484e">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>11 files</summary><table>
<tr>
  <td><strong>browser_download_tests.rs</strong><dd><code>Skip non-Firefox tests on ARM64</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16045/files#diff-359d9c2a125d6fda0f38d3b794fcad796156f1a5786a3ead1ac92759a68a6049">+23/-16</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>browser_tests.rs</strong><dd><code>Skip non-Firefox browser tests on ARM64</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16045/files#diff-c4663bdd0f9bda0b1870b94fe25dc61d2fec5ba6431521a52a330f8002a654c3">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>cache_tests.rs</strong><dd><code>Disable cache tests on ARM64</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16045/files#diff-8f1be668827adf641c7ff682eb2b3f16a7b3c947bb3e7db1f376754ab84c8ff6">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>config_tests.rs</strong><dd><code>Skip non-Firefox config tests on ARM64</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16045/files#diff-336f9940e5c880d6c9d1cfaa5527ac3012bdb9b5515e6994124070297f4d32f6">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>exec_driver_tests.rs</strong><dd><code>Skip non-Firefox driver tests on ARM64</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16045/files#diff-b1fed12daefb8d91e71d88434c68bf3f8f90377c891fed327aa3e36c82d61be0">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>mirror_tests.rs</strong><dd><code>Use Firefox for ARM64 mirror tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16045/files#diff-a43cd18d2fe841d76e1f61be80f4bad1adc886c5c5df398539ae7f6c8272238e">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>offline_tests.rs</strong><dd><code>Use Firefox for ARM64 offline tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16045/files#diff-1841fb63e80f2e98409f9b487aefc50784dfadcf9f2f263db37f1d0d593b4272">+10/-5</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>output_tests.rs</strong><dd><code>Use Firefox for ARM64 output tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16045/files#diff-8843f17a35e855620b36bcb89317ac2f5bdaf288ae9df752478654dc3d41e643">+28/-13</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>proxy_tests.rs</strong><dd><code>Use Firefox for ARM64 proxy tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16045/files#diff-15a6c79463a3e5ec11d3292fd144551f06795ed56434b4f87b5df3abec4d03a3">+10/-4</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>stable_browser_tests.rs</strong><dd><code>Skip non-Firefox stable tests on ARM64</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16045/files#diff-ccfe2de01bf5eee8fbaee613b25e51fecbf4105ce3bb3cac684397c722143280">+6/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>webview_tests.rs</strong><dd><code>Disable webview tests on ARM64</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16045/files#diff-1e8f934eae825effbf1efa8bb5f39567473b2254d8a28bf42c86d25b76073f82">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Enhancement</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>common.rs</strong><dd><code>Add ARM64 detection helper function</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16045/files#diff-e75829e061fe7cd5b9ff65a045c33574cd09617ac149f4ba0114984d2fd14ff7">+6/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>ci-rust.yml</strong><dd><code>Add Ubuntu ARM64 runner to CI</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16045/files#diff-5162db340096b281e8c77ef016f0595e9b7a63e9cb48605a43b685301b4b5ee0">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

dennisameling committed:
[rust] Enable tests in GitHub Actions on Linux arm64

Now that GitHub-hosted Linux arm64 runners are available,
we can start using them to test the Selenium Manager code.

dennisameling committed:
[rust] Selenium Manager: Update tests to work with Linux arm64

Currently, only Firefox and Geckodriver have official support
for Linux arm64. This commit ensures that the Selenium Manager
test suite passes on this platform, by skipping tests on non-
Firefox browsers.

dennisameling committed:
Apply AI review suggestions

cgoldberg commented:
Thanks. I'll leave code review for someone who knows rust.

But I'm a little confused...

> Previously, the code would silently download linux64 binaries which are for Linux x64, and cause segfaults on Linux arm64.

I thought we weren't building Selenium Manager on ARM. Were you running the linux64 binaries through QEMU or something?

Does this PR enable building Selenium Manager for arm64, or just handle the situation when trying to run the linux64 binaries on arm64? If the former, don't we need to update the various bindings and packaging to accommodate this? ... or is this just a step towards enabling Selenium on ARM64?

dennisameling commented:
Sorry for the confusion here.

> I thought we weren't building Selenium Manager on ARM.

Indeed. There have been no Selenium Manager builds for ARM. This PR is a step towards enabling Selenium on ARM64.

The reason I mentioned `linux64` is that if you look at [this](https://github.com/dennisameling/selenium/blob/21e3af98140417fbaead5be3b7f7216257d83d49/rust/src/chrome.rs#L387) code path for example, without any changes for Linux ARM64, the code would just have downloaded the `linux64` Chrome binary because it's in the `else` branch. But Linux ARM64 can't work with that natively, which is why I added [this](https://github.com/SeleniumHQ/selenium/pull/16045/files#diff-5ac161093a229c7c74c75c64f3be31a337c109d1ff11fcb0efe7bcc94d6e77beR387) logic to explicitly tell the user that this scenario isn't supported until Chrome provides native binaries for this platform.

This PR will at least ensure that users of Linux arm64 can use Firefox, which has official binaries for this platform 👍🏼 whenever Edge and Chrome start publishing native binaries too, they can be added through a follow-up PR.

cgoldberg commented:
OK... that makes sense, but it's probably more appropriate to make these changes if/when we start building SM for ARM since these code paths won't be tested or used until then.

dennisameling commented:
With regards to testing, that's why I added https://github.com/SeleniumHQ/selenium/pull/16045/files#diff-5162db340096b281e8c77ef016f0595e9b7a63e9cb48605a43b685301b4b5ee0R38 so that it will start running the tests on native Linux arm64 runners. GitHub [released those in Public Preview in January of this year](https://github.blog/changelog/2025-01-16-linux-arm64-hosted-runners-now-available-for-free-in-public-repositories-public-preview/), and quite some open source projects have started using them already.

I was hoping - by just adding the testing part to CI for now - that the team can at least be confident that the code will work on Linux arm64, and leave the actual builds/publishing for a follow-up PR. But happy to look into that part as well if helpful.

diemol commented:
@bonigarcia, do we need to explicitly build a SM binary for ARM?

bonigarcia commented:
The code looks good to me (thanks a lot for contributing, @dennisameling!).

I don't have an ARM machine to test it. Let's see what CI says about it.

@diemol Yes, we would need to build SM for ARM in Linux. And then, we would need to distribute the new binary together with the other three. Also, the bindings would need to discover the system architecture (in Linux) to use the X64 or ARM SM binary.

dennisameling committed:
Fix formatting

dennisameling committed:
Add release logic for Linux arm64 binaries

dennisameling commented:
Given that all the tests are passing, I just added the necessary logic to publish builds for Linux arm64 as well. I see that [some preparations have already been done](https://github.com/SeleniumHQ/selenium/pull/16052) to start using them in the Python binding.

Is there anything I can help with to push this over the finish line? Thanks! 😊

cgoldberg commented:
@dennisameling 

Thanks! I was the one who did the work to prepare the Python bindings. I want to test building and running Selenium Manager and the Python bindings locally for Linux arm64. Since my only machine is x86, I am going to get a Raspberry Pi next weekend (I'm not even sure if that's the right ARM architecture?)

Hopefully I can get things building and running, and that will give us more confidence to move forward with building for ARM in CI.

Your changes are very helpful and I will update this issue in a week or so once I've had time to play with it.

dennisameling commented:
> I am going to get a Raspberry Pi next weekend (I'm not even sure if that's the right ARM architecture?)

Nice! The Raspberry Pi 3, 4 and 5 are all arm64, so you should be good there. I've been testing natively on an Ubuntu arm64 VM in Parallels on Mac, and can confirm that at least the Rust part is working as expected 👍🏼

cgoldberg commented:
@dennisameling 

I bought a raspberry pi, and everything was surprisingly easy to setup. I was able to build selenium-manager on Raspberry Pi OS (Debian Stable) without problems using our Python source package and `setuptools-rust`. I will continue looking into your PR and update this in a few days.

dennisameling commented on .github/workflows/ci-rust.yml:
```diff
@@ -266,6 +291,7 @@ jobs:
           prerelease: false
           files: |
             artifacts/selenium-manager-linux/selenium-manager-linux
+            artifacts/selenium-manager-linux-arm64/selenium-manager-linux-arm64
```
Realizing I forgot to add `artifacts/selenium-manager-linux-arm64-debug/selenium-manager-linux-arm64-debug.tar` to this list, so let's keep that in mind as we continue working on this PR.

dennisameling commented:
@cgoldberg is there anything I can do to help push this over the finish line? Happy to update/change things if needed. Thanks! 😊

cgoldberg commented:
I should have some time to get back to it soon and I'll let you know if there's anything else you can do. I appreciate your patience 🙂

dennisameling commented:
@cgoldberg quick reminder for this one. Anything I can do to help? Thanks! 🙏🏼

cgoldberg committed:
Merged trunk and fixed conflicts

cgoldberg commented:
@dennisameling

Thanks for your patience on this. There hasn't been much interest from the other maintainers in shipping ARM binaries for Selenium Manager. It's mostly due to perceived low interest/usage on ARM platforms (I don't necessarily agree), lack of driver/browser support on ARM, and some concerns about inflating our package sizes (we include binaries for all platforms in universal packages).

I think we should still go forward trying to get this merged.

cgoldberg commented:
@dennisameling

I checked out your branch, merged trunk, and fixed the merge conflicts and formatting issues.

I committed those changes and pushed them to your branch. Can you make sure my changes look OK?

cgoldberg commented:
@dennisameling

Right now I am getting 5 failures when i run `cargo test` locally.

I'm running on Debian ARM64 13 (Raspberry Pi OS) with the latest rust toolchain. I am building/testing with cargo, not with Bazel, so I don't know if that does anything different?


```
test grid_version_test::case_1 ... FAILED
test grid_version_test::case_3 ... FAILED
test grid_version_test::case_2 ... FAILED
test grid_latest_test ... FAILED
test grid_version_test::case_4 ... FAILED
```

The failures all look like:

```
command=`"/home/cgoldberg/code/selenium/rust/target/debug/selenium-manager" "--grid" "nightly" "--output" "json"`
code=65
stdout=
{
  \"logs\": [
    {
      \"level\": \"ERROR\",
      \"timestamp\": 1764722307,
      \"message\": \"Invalid cross-device link (os error 18)\"
    }
  ],
  \"result\": {
    \"code\": 65,
    \"message\": \"Invalid cross-device link (os error 18)\",
    \"driver_path\": \"\",
    \"browser_path\": \"\"
  }
}
```

I also built selenium-manager from trunk branch, and I get the same failures:

```
$ selenium-manager --grid 4.36.0
[2025-12-03T00:42:21.022Z ERROR] Invalid cross-device link (os error 18)
```

... so it's not something your code introduced, but it is something we need to investigate and fix.

Do you have any idea what those failures mean, and can you reproduce them?

cgoldberg committed:
Revert test fixes and fix formatting

cgoldberg committed:
Merge branch 'trunk' into add-selenium-manager-linux-arm64

cgoldberg commented:
Also, when I run selenium-manager on arm64, it is downloading the x86-64 version of Firefox, not the ARM version (but it downloads the correct geckodriver for this architecture).

cgoldberg committed:
Merge branch 'trunk' into add-selenium-manager-linux-arm64

cgoldberg commented:
@dennisameling 


I have your branch updated from upstream trunk, but I am running into some issues. Do you have any time to look at this?

I'm testing on a Raspberry Pi 5 running Raspberry Pi OS (Debian 13) with the latest Rust toolchain installed.

----

When I run the selenium Rust test suite in your branch, I get 5 failures:

```
failures:

---- grid_version_test::case_2 stdout ----

thread 'grid_version_test::case_2' (71678) panicked at /rustc/ded5c06cf21d2b93bffd5d884aa6e96934ee4234/library/core/src/ops/function.rs:250:5:
Unexpected failure.
code=65
stderr=""
command=`"/home/cgoldberg/code/selenium/rust/target/debug/selenium-manager" "--grid" "4.9.0" "--output" "json"`
code=65
stdout=
{
  \"logs\": [
    {
      \"level\": \"ERROR\",
      \"timestamp\": 1766289434,
      \"message\": \"Invalid cross-device link (os error 18)\"
    }
  ],
  \"result\": {
    \"code\": 65,
    \"message\": \"Invalid cross-device link (os error 18)\",
    \"driver_path\": \"\",
    \"browser_path\": \"\"
  }
}

stderr=""

note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- grid_version_test::case_1 stdout ----

thread 'grid_version_test::case_1' (71628) panicked at /rustc/ded5c06cf21d2b93bffd5d884aa6e96934ee4234/library/core/src/ops/function.rs:250:5:
Unexpected failure.
code=65
stderr=""
command=`"/home/cgoldberg/code/selenium/rust/target/debug/selenium-manager" "--grid" "4.8.0" "--output" "json"`
code=65
stdout=```
{
  \"logs\": [
    {
      \"level\": \"ERROR\",
      \"timestamp\": 1766289434,
      \"message\": \"Invalid cross-device link (os error 18)\"
    }
  ],
  \"result\": {
    \"code\": 65,
    \"message\": \"Invalid cross-device link (os error 18)\",
    \"driver_path\": \"\",
    \"browser_path\": \"\"
  }
}

stderr=""


---- grid_latest_test stdout ----

thread 'grid_latest_test' (71627) panicked at /rustc/ded5c06cf21d2b93bffd5d884aa6e96934ee4234/library/core/src/ops/function.rs:250:5:
Unexpected failure.
code=65
stderr=""
command=`"/home/cgoldberg/code/selenium/rust/target/debug/selenium-manager" "--grid" "--output" "json"`
code=65
stdout=
{
  \"logs\": [
    {
      \"level\": \"ERROR\",
      \"timestamp\": 1766289434,
      \"message\": \"Invalid cross-device link (os error 18)\"
    }
  ],
  \"result\": {
    \"code\": 65,
    \"message\": \"Invalid cross-device link (os error 18)\",
    \"driver_path\": \"\",
    \"browser_path\": \"\"
  }
}

stderr=""


---- grid_version_test::case_3 stdout ----

thread 'grid_version_test::case_3' (71691) panicked at /rustc/ded5c06cf21d2b93bffd5d884aa6e96934ee4234/library/core/src/ops/function.rs:250:5:
Unexpected failure.
code=65
stderr=```""```
command=`"/home/cgoldberg/code/selenium/rust/target/debug/selenium-manager" "--grid" "4.10.0" "--output" "json"`
code=65
stdout=
{
  \"logs\": [
    {
      \"level\": \"ERROR\",
      \"timestamp\": 1766289434,
      \"message\": \"Invalid cross-device link (os error 18)\"
    }
  ],
  \"result\": {
    \"code\": 65,
    \"message\": \"Invalid cross-device link (os error 18)\",
    \"driver_path\": \"\",
    \"browser_path\": \"\"
  }
}

stderr=""


---- grid_version_test::case_4 stdout ----

thread 'grid_version_test::case_4' (71710) panicked at /rustc/ded5c06cf21d2b93bffd5d884aa6e96934ee4234/library/core/src/ops/function.rs:250:5:
Unexpected failure.
code=65
stderr=""
command=`"/home/cgoldberg/code/selenium/rust/target/debug/selenium-manager" "--grid" "nightly" "--output" "json"`
code=65
stdout=
{
  \"logs\": [
    {
      \"level\": \"ERROR\",
      \"timestamp\": 1766289441,
      \"message\": \"Invalid cross-device link (os error 18)\"
    }
  ],
  \"result\": {
    \"code\": 65,
    \"message\": \"Invalid cross-device link (os error 18)\",
    \"driver_path\": \"\",
    \"browser_path\": \"\"
  }
}

stderr=""



failures:
    grid_latest_test
    grid_version_test::case_1
    grid_version_test::case_2
    grid_version_test::case_3
    grid_version_test::case_4
```

----

I also did some other testing using the `selenium-manager` binary I built from your branch...

Trying to download edge/msedgedriver gives me the expected error.

```
$ ./selenium-manager --clear-cache --force-browser-download --arch=arm64 --os=linux --browser=edge
[2025-12-21T04:10:16.646Z ERROR] Linux arm64 is not supported yet by Microsoft Edge. Please try another browser.
```

Trying to download chrome/chromedriver should give me the error `"Linux arm64 is not supported yet by Google Chrome. Please try another browser."`... but instead it downloads the x64 version of chrome and chromedriver.

```
$ ./selenium-manager --clear-cache --force-browser-download --arch=arm64 --os=linux --browser=chrome
[2025-12-21T04:09:08.989Z INFO ] Driver path: /home/cgoldberg/.cache/selenium/chromedriver/linux64/143.0.7499.169/chromedriver
[2025-12-21T04:09:08.989Z INFO ] Browser path: /home/cgoldberg/.cache/selenium/chrome/linux64/143.0.7499.169/chrome
```

Downloading firefox and geckodriver should work. It downloads the correct arm64 version of geckodriver, but it downloads the x64 version of Firefox. Intead of the `linux64` version, it should have downloaded the latest `linux-aarch64` version from https://ftp.mozilla.org/pub/firefox/releases/146.0.1/linux-aarch64/en-US/firefox-146.0.1.tar.xz

```
$ ./selenium-manager --clear-cache --force-browser-download --arch=arm64 --os=linux --browser=firefox
[2025-12-21T04:06:12.039Z INFO ] Driver path: /home/cgoldberg/.cache/selenium/geckodriver/linux-arm64/0.36.0/geckodriver
[2025-12-21T04:06:12.039Z INFO ] Browser path: /home/cgoldberg/.cache/selenium/firefox/linux64/146.0.1/firefox
```

cgoldberg committed:
Merge branch 'trunk' into add-selenium-manager-linux-arm64

cgoldberg committed:
Merge branch 'trunk' into add-selenium-manager-linux-arm64

cgoldberg committed:
Merge branch 'trunk' into add-selenium-manager-linux-arm64

cgoldberg commented:
The `Invalid cross-device link (os error 18)` errors were fixed in another PR and no longer occur.

cgoldberg committed:
Merge branch 'trunk' into add-selenium-manager-linux-arm64

dennisameling committed:
Fix Firefox Linux arm64 logic and fail early on Chrome/Edge Linux arm64

dennisameling committed:
Merge remote-tracking branch 'upstream/trunk' into add-selenium-manager-linux-arm64

dennisameling committed:
Add missing debug tar to CI

dennisameling commented:
@cgoldberg apologies for the silence on my side - I had many other things going on and didn't have bandwidth to look into this. Thank you for testing this so thoroughly!

I just addressed your comments, and also noticed that Edge was actually downloaded for x64 _before_ throwing that error. I just improved the logic so that the error is thrown before `selenium-manager` even tries to download the binaries.

```
$ ./selenium-manager --clear-cache --force-browser-download --arch=arm64 --os=linux --browser=edge
[2026-02-02T17:47:53.301Z ERROR] Linux arm64 is not supported yet by edge. Please try another browser.

$ ./selenium-manager --clear-cache --force-browser-download --arch=arm64 --os=linux --browser=chrome
[2026-02-02T17:49:17.867Z ERROR] Linux arm64 is not supported yet by chrome. Please try another browser.

$ ./selenium-manager --clear-cache --force-browser-download --arch=arm64 --os=linux --browser=firefox
[2026-02-02T17:49:51.077Z INFO ] Driver path: /usr/bin/geckodriver
[2026-02-02T17:49:51.077Z INFO ] Browser path: /home/parallels/.cache/selenium/firefox/linux-arm64/147.0.2/firefox
```

Would you mind testing on your end again? Thanks!

dennisameling committed:
Enable cache tests

cgoldberg commented:
@dennisameling I'll take a look.. thanks!

cgoldberg committed:
Merge branch 'trunk' into add-selenium-manager-linux-arm64

dennisameling committed:
Revert "Enable cache tests"

This reverts commit 1384996d9b0ca44c9a723c0f46c14e910df58989.

cgoldberg committed:
Merge branch 'trunk' into add-selenium-manager-linux-arm64

dennisameling committed:
Merge remote-tracking branch 'upstream/trunk' into add-selenium-manager-linux-arm64

titusfortner commented:
@dennisameling thanks for the PR (sorry for all the churn on it the past few days).

I reworked how Firefox on linux-arm64 is gated. Instead of the inline major_browser_version < 136 check plus the driver-side minor_driver_version ("as of 0.32") branch, availability is now enforced through the standard minimum-browser-version download gate (136 for linux-aarch64). Because that floor guarantees the resolved geckodriver is always ≥ 0.32 (first version with aarch64 builds), the minor_driver_version special-casing is no longer reachable and was removed. Simpler, and the unavailable-download error now includes the architecture.

This PR just builds and stages the selenium manager for Linux ARM without any of the bazel wiring to make it available to the bindings or to release with the bindings. We know we're going to have this once #17586 is sorted, but since it is stuck I think we're good to merge this now. We'll need to get the bazel wiring and bindings implementations sorted next.

titusfortner commented:
To close the loop this is successfully generating binaries in our artifacts repo as desired: https://github.com/SeleniumHQ/selenium_manager_artifacts/releases/tag/selenium-manager-25d4fd9

cgoldberg commented:
@dennisameling thanks for the work on this
@titusfortner thanks for getting it over the finish line!

mkurz commented:
Does that mean selenium will soon ship with Selenium Manager for Linux arm64?

cgoldberg commented:
@mkurz work needs to be done to integrate and package it with every language binding. I don't know the current plan for that, but this is a big step closer.
